test/utils/cryptography/P256.t.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {Test} from "forge-std/Test.sol";
+import {P256} from "@openzeppelin/contracts/utils/cryptography/P256.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+
+contract P256Test is Test {
```

Comment:
think we should have a unit test for out of bound public keys -- i can add one

Reply:
it also doesn't hurt to have unit tests that import and use https://github.com/C2SP/wycheproof

Reply:
> it also doesn't hurt to have unit tests that import and use https://github.com/C2SP/wycheproof

Agree, I was thinking in loading the test vectors as we did in RSA (but for hardhat), wdyt?

Reply:
added with foundry here: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4881/commits/93629369e57325c87463989994415106066acdd1 

checked with hardhat and is pretty slow compared to this, but we can do it if we really want to

Reply:
I see the tests are failing. Previously I shared these cases I identified where failing with @Amxx and I think these are those failing here too. FYI 

#### Case 1 (OZ = false, P256Verifier = true)
r:  0x000000000000000000000000000000004319055358e8617b0c46353d039cdaab
s:  0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc63254e
x:  0x0ad99500288d466940031d72a9f5445a4d43784640855bf0a69874d2de5fe103
> wycheproof/ecdsa_secp256r1_sha256_p1363_test.json EcdsaP1363Verify SHA-256 #113: k*G has a large x-coordinate
> wycheproof/ecdsa_webcrypto_test.json EcdsaP1363Verify SHA-256 #113: k*G has a large x-coordinate

#### Case 2 (OZ = false, P256Verifier = true)
r:  0x7cf27b188d034f7e8a52380304b51ac3c08969e277f21b35a60b48fc47669978
s:  0xb6db6db6249249254924924924924924625bd7a09bec4ca81bcdd9f8fd6b63cc
x:  0xc6a771527024227792170a6f8eee735bf32b7f98af669ead299802e32d7c3107
> wycheproof/ecdsa_secp256r1_sha256_test.json EcdsaVerify SHA-256 #345: extreme value for k and s^-1
> wycheproof/ecdsa_secp256r1_sha256_p1363_test.json EcdsaP1363Verify SHA-256 #177: extreme value for k and s^-1

#### Case 3 (OZ = false, P256Verifier = true)
r:  0x7cf27b188d034f7e8a52380304b51ac3c08969e277f21b35a60b48fc47669978
s:  0xb6db6db6249249254924924924924924625bd7a09bec4ca81bcdd9f8fd6b63cc
x:  0xc6a771527024227792170a6f8eee735bf32b7f98af669ead299802e32d7c3107
> wycheproof/ecdsa_secp256r1_sha256_test.json EcdsaVerify SHA-256 #345: extreme value for k and s^-1
> wycheproof/ecdsa_secp256r1_sha256_p1363_test.json EcdsaP1363Verify SHA-256 #177: extreme value for k and s^-1

#### Case 4 (OZ = false, P256Verifier = true)
r:  0x000000000000000000000000000000004319055358e8617b0c46353d039cdaab
s:  0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc63254e
x:  0xd705d16f80987e2d9b1a6957d29ce22febf7d10fa515153182415c8361baaca4
> wycheproof/ecdsa_webcrypto_test.json EcdsaP1363Verify SHA-256 #113: k*G has a large x-coordinate
> wycheproof/ecdsa_secp256r1_sha256_p1363_test.json EcdsaP1363Verify SHA-256 #113: k*G has a large x-coordinate

Reply:
FYI^ the `s` parameters are above the malleability threshold so we have to make sure we get to the underlying issue not just skip in case we add malleability checks.

Reply:
also want to make sure we are good for the following cases that occurred in other well known implementations:
- [verify signatures whose R,S > curve order](https://github.com/paritytech/libsecp256k1/commit/b525d5d318d9672a40250c1725fa1bb3156688b7)
- [all zero signatures](https://github.com/paulmillr/noble-secp256k1/commit/9082b405a3d1958b5b6a699bb408215ae30eea5b#diff-177d97aec07fd5b22f04049b34d74939b1d8364a7a83e8fb69ee3ee8695105fbR256-R270)
- [invalid r,s](https://github.com/paulmillr/noble-secp256k1/commit/c514f79ebdab1042dac214446f1ca32214a5edfd#diff-177d97aec07fd5b22f04049b34d74939b1d8364a7a83e8fb69ee3ee8695105fbR242-R255)
- [scalar multiplications](https://github.com/ziglang/zig/issues/15267)
- [message is curve order](https://github.com/paritytech/libsecp256k1/issues/62) -- verify instead of sign.

differential fuzzing with [cryptofuzz](https://github.com/guidovranken/cryptofuzz) would be amazing, will check it out.

Reply:
> FYI^ the s parameters are above the malleability threshold so we have to make sure we get to the underlying issue not just skip in case we add malleability checks.

Yes, found out the hard way trying to debug a panic lol

> [verify signatures whose R,S > curve order](https://github.com/paritytech/libsecp256k1/commit/b525d5d318d9672a40250c1725fa1bb3156688b7)

Seems like these are already covered by the test vectors and fixed by bounding r to N